test/test_dataloader2.py

```diff
@@ -52,6 +56,33 @@ def test_dataloader2(self) -> None:
             self.assertEqual(batch, expected_batch)
             expected_batch += 1
 
+    def test_dataloader2_len(self) -> None:
+
+        reading_services = (
+            None,
+            MultiProcessingReadingService(num_workers=2),
+            PrototypeMultiProcessingReadingService(num_workers=2),
+        )
+
+        sharding_dp = IterableWrapper(range(10)).sharding_filter()
+        filter_dp = IterableWrapper(range(10)).filter(_filter_fn)
+
+        for rs in reading_services:
+            # Functional Test: Case with sharding
+            data_loader_sharding: DataLoader2 = DataLoader2(datapipe=sharding_dp, reading_service=rs)
+            # Call `__len__` before initialization
+            self.assertEqual(10, len(data_loader_sharding))
+            _it = iter(data_loader_sharding)
+            # Length should stay the same after calling `__iter__`
+            self.assertEqual(10, len(data_loader_sharding))
+            list(_it)
+            self.assertEqual(10, len(data_loader_sharding))
+
+            # Functional Test: Case with filter
+            data_loader_filter: DataLoader2 = DataLoader2(datapipe=filter_dp, reading_service=rs)
```

Comment:
Since `rs` is being re-used, this would've failed without the previous PR that deepcopies.